### PR TITLE
Caching setting values

### PR DESCRIPTION
**Pull request by JapSeyz**

### CODE REVIEW

**Comment by handiwijoyo** [comment]
@JapSeyz Can you provide the details why do we need this?

**Comment by JapSeyz** [comment]
@handiwijoyo Caching the setting to avoid hitting the DB on every page-load for the setting.

Same thing could be done for menu.

I'll admit, it's not the biggest boost, but for API rutes that needs a single setting value, and can be called a few hundred times a minute at a peak hour, calling the DB for the setting-value every time is redundant if it can be stored in Redis.

It doesn't give you any noticable difference on normal view-based routes, but it can really add up on API routes.

**Comment by handiwijoyo** [comment]
@JapSeyz Ok. The idea looks fine for me. But I think we need to add events for settings controller. Then we can put the cache inside the listener and have a config option to enable and disable the cache.

**Comment by handiwijoyo** [comment]
After take a look into the Voyager::setting() https://github.com/the-control-group/voyager/blob/1.0/src/Voyager.php#L161

image

It's already have different cache method. We might need to change this too.

**Comment by JapSeyz** [comment]
@handiwijoyo That cache is only request-wide. Once the request is over, the caching is removed.

I'll look into implementing the global cache there too, and I'll update the PR during the weekend.

I'll also look into the events in the weekend.

**Comment by JapSeyz** [comment]
@handiwijoyo I have added a listener and moved the code to the Voyager.php file.

If you agree with what's done, I'll fix the StyleCI issues

File: src/Events/SettingUpdated.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
```
Comment (fletch3555) [inline src/Events/SettingUpdated.php]: Why is this added here? We don't have it elsewhere in the project.  It may be a good idea, I don't know, but I'm a bigger fan of consistency.  Please remove it from this and other files.

Comment (JapSeyz) [reply src/Events/SettingUpdated.php]: I will, it's one of the things from my php Cs fixer.

Once one of you approve that the actual code is alright I'll give it a once over and remove any styleCI issues as well as these. I think I also wrote it in the comment above.

File: src/Voyager.php [hunk 2]
```diff
@@ -160,10 +160,27 @@ public function formFields()
 
     public function setting($key, $default = null)
     {
+        $globalCache = config('voyager.settings.cache', false);
+
+        if ($globalCache && Cache::tags('settings')->has($key)) {
+            return Cache::tags('settings')->get($key);
+        }
+
         if ($this->setting_cache === null) {
```
Comment (fletch3555) [inline src/Voyager.php:218]: Was it intentional to leave the per-request/local cache in place?  I feel like this would be much simpler if it was simply replaced with `Cache::` calls like you've already done.

**Comment by fletch3555** [comment]
@JapSeyz, sorry for the delay in review here.  Looks good.  Only question is stated above, why the multiple layers of cache?

File: src/Voyager.php (hunk 2 above)
Comment (JapSeyz) [reply src/Voyager.php:218]: It's been a while, but if I recall correctly, it was to disallow users to use the caching function, if they for some reason wanted to do that. It can be stripped out without issues

File: src/Listeners/ClearCachedSettingValue.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Listeners;
```
Comment (fletch3555) [inline src/Listeners/ClearCachedSettingValue.php]: TravisCI is failing due to this line.  Please change it to the correct namespace (`TCG\Voyager\Listeners`)

File: src/Helpers/helpers.php [hunk 4]
```diff
@@ -1,20 +1,22 @@
 <?php
 
-if (!function_exists('setting')) {
+declare(strict_types=1);
+
+if (!\function_exists('setting')) {
```
Comment (fletch3555) [inline src/Helpers/helpers.php]: None of the changes in this file are necessary for this PR.  Please revert them.

File: src/Events/SettingUpdated.php (hunk 1 above)
Comment (fletch3555) [reply src/Events/SettingUpdated.php]: Sorry, I know this has been a while, but if you don't mind cleaning up the phpcs changes, I'm good with it.

I'm not too worried about StyleCI, since that will fix itself.

Thanks!

**Review by fletch3555 — CHANGES_REQUESTED** [review]

**Review by fletch3555 — APPROVED** [review]
